IGNITE-23383 Doc for transactional aware queries (#13116)
apache/ignite: docs/_docs/SQL/sql-calcite.adoc

----
SELECT /*+ DISABLE_RULE('MergeJoinConverter') */ T1.* FROM TBL1 T1 JOIN TBL2 T2 ON T1.V1=T2.V1 WHERE T2.V2=?
----

== Transactions [[transactions]]

> Transactions on `READ_COMMITTED` isolation level are supported for SQL queries in Calcite query engine.

Reviewer: ```suggestion
The Calcite-based query engine supports SQL transactions with READ_COMMITTED isolation.
```

Reviewer: PESSIMISTIC, READ_COMMITTED !! Not Optimistic
I think it also need to be mention that some kind of Frankenstein is currently implemented, i.e : 
```
        try (Transaction tx = client.transactions().txStart(PESSIMISTIC, READ_COMMITTED)) {
            sql("some SQL ...
```
and not like:
```
BEGIN TRANSACTION;
  UPDATE accounts SET balance = balance - 500 WHERE id = 1;
  UPDATE accounts SET balance = balance + 500 WHERE id = 2;
COMMIT;
```

To ensure backward compatibility transactions support disabled, by default.

> To enable, please, configure like the following:

Reviewer: ```suggestion
To ensure backward compatibility, transaction support is disabled by default. To enable it, use the following configuration:
```

[tabs]
--
tab:XML[]
[source,xml]
----

> <bean class="org.apache.ignite.configuration.IgniteConfiguration">

Reviewer: what about java code configuration ?

Author: I'm sure any Ignite user can relect XML config to the java one and vice versa :)

Reviewer: it`s just about friendy UI ... and addressd to newbie guys too

    <property name="transactionConfiguration">
        <bean class="org.apache.ignite.configuration.TransactionConfiguration">
            <property name="txAwareQueriesEnabled" value="true" />
            ...
        </bean>
    </property>
    ...
</bean>
----
--

The following API "transactional aware":

* key-value API.
* SQL queries with Calcite engine.
* Scan queries.

"Transactional aware" means:

1. ACID in the same way as key-value API.

> 2. Any data changed or removed by the query will not be visible to the concurrent query until transaction committed.

Reviewer: ```suggestion
2. Any data modified or deleted within a transaction will not be visible to concurrent transactions until it is committed.
```

> 3. Any data changed by the query will be returned updated by subsequent queries of the same transaction.

Reviewer: don`t understand at all (

Author: Please, see example:

```
try (Transaction tx = ignite.transactions.start()) {
    cache.put(1, "test");
    var res = sql("SELECT _VAL FROM cache WHERE _KEY = 1");
}
```

If `txAwareQueriesEnabled == false' SELECT will return empty result, because, transaction not committed.
But, if `txAwareQueriesEnabled == true' SELECT will return "test" row.

Reviewer: I mean: it`s not human readable  ))) 
do you agree to change like : 
Data modified by a query will be returned with updated values by subsequent queries within the same transaction.

> 4. And data removed by the query will not be returned by subsequent queries of the same transaction.

Reviewer: ```suggestion
4. Any data removed by the query will not be returned by subsequent queries of the same transaction.
```

may be :
Data deleted within a transaction will not be returned by subsequent queries in that same transaction.

> So, when `txAwareQueriesEnabled = true` you can:

Reviewer: i\`d prefer not 1-st form, not "you" but smth like: With `txAwareQueriesEnabled = true` become possible ... 
Or 
Setting txAwareQueriesEnabled = true enables the use of [feature]

* Perform INSERT, UPDATE, DELETE statements with regular transaction guarantees.
* Mix key-value, SQL and Scan queries to process same dataset.

[NOTE]
====
[discrete]
=== SELECT ... FOR UPDATE
Currently, no locks are held by the UPDATE or SELECT statements.

> `SELECT ... FOR UPDATE` statement not supported.

Reviewer: Let`s describe syntax like :
`update t set val = val + 1 WHERE id = 1`
what does it means : Currently, no locks are held by the UPDATE or SELECT

lets mean:
For example, if you execute a query like `SET salary = salary + 50 WHERE id = 1` concurrently across multiple threads, note that due to the lost-update anomaly, the final value may not equal the original salary plus 50 multiplied by the number of threads.

or 

When running UPDATE ... SET salary = salary + 50 WHERE id = 1concurrently from multiple threads, be aware that the lost-update anomaly may cause the final value to differ frominitial_salary + (50 × thread_count).

> This mean, lost update are possible with transactional SQL queries in case several transactions modifies one key concurrently.

Reviewer: update**s** are ot update **is**

Reviewer: This means lost updates can occur when multiple transactions concurrently modify the same key.
